testutil/wait.go

```diff
@@ -138,10 +138,27 @@ type rpcFn func(string, interface{}, interface{}) error
 func WaitForLeader(t testing.TB, rpc rpcFn) {
 	t.Helper()
 	WaitForResult(func() (bool, error) {
-		args := &structs.GenericRequest{}
+		args := &structs.GenericRequest{
+			QueryOptions: structs.QueryOptions{
+				Namespace: "default",
+				Region:    "global",
+			},
+		}
 		var leader string
 		err := rpc("Status.Leader", args, &leader)
-		return leader != "", err
+		if err != nil {
+			return leader != "", err
+		}
+		if leader == "" {
+			return false, err
+		}
+
+		// As of 1.7 the RSA key generation slows down keyring initialization so
+		// much that tests relying on the keyring being initialized fail unless
+		// they wait until the key was generated
+		var resp structs.KeyringListPublicResponse
+		err = rpc("Keyring.ListPublic", args, &resp)
+		return err == nil, err
```

Comment:
**Oof.** RSA key generation is _so much slower than ed25519_ that it causes heaps of tests to fail. Anything that expects the keyring to be initialized but doesn't explicitly wait for that to occur fails with a `keyring has not been initialized yet` error.

I wrote a [quick benchmark for comparing rsa.GenerateKey vs ed25519.NewKeyFromSeed](https://gist.github.com/schmichael/a3e66d72de128cff273dbe65cdca3e77) and got:

```
~/src/benchcrypto$ go test -bench .
goos: linux
goarch: amd64
pkg: github.com/schmichael/benchcrypto
cpu: 12th Gen Intel(R) Core(TM) i9-12900H
BenchmarkRSA-20               16          63629942 ns/op
BenchmarkEd25519-20       109570             10977 ns/op
PASS
ok      github.com/schmichael/benchcrypto       3.883s
```

That's **63ms** to generate an RSA key vs 11 microseconds for ed25519.

Generating a new RSA key only occurs every 30 days, so the latency in production environments doesn't matter.

However this is a pretty big hassle for tests. Technically this was a race condition just waiting to cause flakiness, so maybe I shouldn't feel too bad?

Reply:
Would it make sense to change something like `TestConfigForServer` to use `Ed25519` by default since I think most tests don't care about the exact algorithm being used?

Or maybe preload test servers with hardcoded RSA material so they don't have to be computed all the time?

Reply:
I'm exploring the `testutil.WaitForKeyring` route first as the fact that our tests have hidden races in them is is going to bother me regardless. :sweat_smile: 

Luiz's options are good fallbacks (or could even be nice optimizations in addition to WaitForKeyring), but all fall into the category of widening the differences between test servers and reality. For 99.9% of tests that extra difference isn't worth worrying about, so I don't think it's a big concern.

Thanks for helping me brainstorm this!